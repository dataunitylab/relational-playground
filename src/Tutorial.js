// @flow
import Joyride, {STATUS} from 'react-joyride';
import React, {useState} from 'react';
import {useCookies} from 'react-cookie';

import type {StatelessFunctionalComponent} from 'react';

type CallBackProps = {
  status: string,
};

const redoText = 'Redo tutorial';
const stepStyle = {tooltipContainer: {textAlign: 'left'}};

/** Container for all components of the tutorial */
const Tutorial: StatelessFunctionalComponent<{}> = (props) => {
  let initialText, shouldRun;
  const [cookies, setCookie] = useCookies(['tutorial']);
  if (cookies.tutorial === undefined) {
    initialText = 'Tutorial';
    shouldRun = true;
    setCookie('tutorial', 'true', {path: '/', sameSite: 'strict'});
  } else {
    initialText = redoText;
    shouldRun = false;
  }

  const [buttonText, setButtonText] = useState(initialText);
  const [runTutorial, setRunTutorial] = useState(shouldRun);

  const handleClickStart = (e: SyntheticMouseEvent<HTMLElement>) => {
    e.preventDefault();
    setRunTutorial(true);
  };

  const handleJoyrideCallback = (data: CallBackProps) => {
    const {status} = data;
    const finishedStatuses: string[] = [STATUS.FINISHED, STATUS.SKIPPED];

    if (finishedStatuses.includes(status)) {
      setButtonText(redoText);
      setRunTutorial(false);
    }
  };

  const steps = [
    {
      content: <h2>Welcome to Relational Playground!</h2>,
      locale: {skip: <strong aria-label="skip">Skip tutorial</strong>},
      placement: 'center',
      target: 'body',
    },
    {
      content: (
        <div>
          <h2>Enter SQL queries here</h2>
          <p>
            Note that not all SQL queries are supported (e.g. aggregation,
            arithmetic, and function calls).
          </p>
        </div>
      ),
      spotlightPadding: 4,
      target: '.SqlEditor',
      styles: stepStyle,
    },
    {
      content: (
        <div>
          <h4>Source relations are displayed here</h4>,
          <p>Any of these relations can be used in your SQL queries.</p>
        </div>
      ),
      spotlightPadding: 4,
      target: '.sourceTableContainer',
      styles: stepStyle,
    },
    {
      content: (
        <div>
          <h4>Relational algebra expressions</h4>
          <p>
            The expression is generated from the SQL queries entered above. The
            expression will automatically be updated based on the query. You can
            select any subexpression to view the data the relation contains
            below.
          </p>
        </div>
      ),
      spotlightPadding: 4,
      target: '.relExprContainer',
      styles: stepStyle,
    },
    {
      content: (
        <div>
          <h4>Relation algebra expressions data</h4>
          <p>
            When you select an expression above, the data contained in that
            expression will be displayed here. You can select the top-level
            expression to see the final results of the query or any intermediate
            expression to see what relations result from each step.
          </p>
        </div>
      ),
      spotlightPadding: 4,
      target: '.dataContainer',
      styles: stepStyle,
    },
    {
      content: (
        <div>
          <h4>Tree view</h4>
          <p>
            Toggling this checkbox will enable a tree view for relational
            algebra expressions. You can switch back to a linear view by
            unchecking the box.
          </p>
        </div>
      ),
      spotlightPadding: 4,
      target: '.relExprContainer .toggle',
      styles: stepStyle,
    },
  ];

  return (
    <div>
      <Joyride
        callback={handleJoyrideCallback}
        continuous
        scrollToFirstStep
        showProgress
        showSkipButton
        styles={{
          options: {
            zIndex: 10000,
          },
        }}
        run={runTutorial}
        steps={steps}
      />
      <button className="button" onClick={handleClickStart}>
        {buttonText}
      </button>
    </div>
  );
};

export default Tutorial;
