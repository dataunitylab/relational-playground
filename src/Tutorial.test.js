import React from 'react';
import Tutorial from './Tutorial';
import Cookies from 'universal-cookie/cjs';
import {shallow} from 'enzyme';
import JoyrideTooltipContainer from 'react-joyride';

/** @test {Tutorial} */
it('correctly renders tutorial given no cookie', () => {
  let cookies = new Cookies();
  const wrapper = shallow(<Tutorial />);
  let tooltip = wrapper.find(JoyrideTooltipContainer);

  // expect(tooltip).toBeDefined();
  expect(tooltip).toMatchSnapshot();
});

/** @test {Tutorial} */
it('cookie exists, no tutorial', () => {
  let cookies = new Cookies();
  cookies.set('tutorial', 'true', {path: '/'});

  const wrapper = shallow(<Tutorial />);
  let button = wrapper.find('.button').text();

  expect(button).toContain('Redo tutorial');
});
