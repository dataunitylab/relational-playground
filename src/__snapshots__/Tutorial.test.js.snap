// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`correctly renders tutorial given no cookie 1`] = `
<Joyride
  callback={[Function]}
  continuous={true}
  debug={false}
  disableCloseOnEsc={false}
  disableOverlay={false}
  disableOverlayClose={false}
  disableScrollParentFix={false}
  disableScrolling={false}
  hideBackButton={false}
  run={true}
  scrollDuration={300}
  scrollOffset={20}
  scrollToFirstStep={true}
  showProgress={true}
  showSkipButton={true}
  spotlightClicks={false}
  spotlightPadding={10}
  steps={
    Array [
      Object {
        "content": <h2>
          Welcome to Relational Playground!
        </h2>,
        "locale": Object {
          "skip": <strong
            aria-label="skip"
          >
            Skip tutorial
          </strong>,
        },
        "placement": "center",
        "target": "body",
      },
      Object {
        "content": <div>
          <h2>
            Enter SQL queries here
          </h2>
          <p>
            Note that not all SQL queries are supported (e.g. aggregation, arithmetic, and function calls).
          </p>
        </div>,
        "spotlightPadding": 4,
        "styles": Object {
          "tooltipContainer": Object {
            "textAlign": "left",
          },
        },
        "target": ".SqlEditor",
      },
      Object {
        "content": <div>
          <h4>
            Source relations are displayed here
          </h4>
          ,
          <p>
            Any of these relations can be used in your SQL queries.
          </p>
        </div>,
        "spotlightPadding": 4,
        "styles": Object {
          "tooltipContainer": Object {
            "textAlign": "left",
          },
        },
        "target": ".sourceTableContainer",
      },
      Object {
        "content": <div>
          <h4>
            Relational algebra expressions
          </h4>
          <p>
            The expression is generated from the SQL queries entered above. The expression will automatically be updated based on the query. You can select any subexpression to view the data the relation contains below.
          </p>
        </div>,
        "spotlightPadding": 4,
        "styles": Object {
          "tooltipContainer": Object {
            "textAlign": "left",
          },
        },
        "target": ".relExprContainer",
      },
      Object {
        "content": <div>
          <h4>
            Relation algebra expressions data
          </h4>
          <p>
            When you select an expression above, the data contained in that expression will be displayed here. You can select the top-level expression to see the final results of the query or any intermediate expression to see what relations result from each step.
          </p>
        </div>,
        "spotlightPadding": 4,
        "styles": Object {
          "tooltipContainer": Object {
            "textAlign": "left",
          },
        },
        "target": ".dataContainer",
      },
      Object {
        "content": <div>
          <h4>
            Tree view
          </h4>
          <p>
            Toggling this checkbox will enable a tree view for relational algebra expressions. You can switch back to a linear view by unchecking the box.
          </p>
        </div>,
        "spotlightPadding": 4,
        "styles": Object {
          "tooltipContainer": Object {
            "textAlign": "left",
          },
        },
        "target": ".relExprContainer .toggle",
      },
    ]
  }
  styles={
    Object {
      "options": Object {
        "zIndex": 10000,
      },
    }
  }
/>
`;
